src/jit/compiler.h

```diff
@@ -297,6 +297,208 @@ enum RefCountState
+void dumpVariableLiveRange(const VariableLiveRange* varLiveRange, emitter* _emitter, const CodeGenInterface* codeGen);
+
+//--------------------------------------------
+//
+// LiveRangeDumper: Used for debugging purposes during code
+//  generation on genCodeForBBList. Keeps an iterator to the fist
+//  edited/added "VariableLiveRange" of a variable during the
+//  generation of code of one block.
+//
+// Notes:
+//  The first "VariableLiveRange" reported for a variable during
+//  a BasicBlock is sent to "setDumperStartAt" so we can dump all
+//  the "VariableLiveRange"s from that one.
+//  After we dump all the "VariableLiveRange"s we call "reset" with
+//  the "liveRangeList" to set the barrier to nullptr or the last
+//  "VariableLiveRange" if it is opened.
+//  If no "VariableLiveRange" was edited/added during block,
+//  the iterator points to the end of variable's LiveRangeList.
+//
+class LiveRangeDumper
+{
+private:
+    LiveRangeListIterator m_StartingLiveRange; // Iterator to the first edited/added position
+                                               // during actual block code generation. If last
+                                               // block had a closed "VariableLiveRange" (with
+                                               // a valid "m_EndEmitLocation") and not changes
+                                               // were applied to variable liveness, it points
+                                               // to the end of variable's LiveRangeList.
+    bool m_hasLiveRangestoDump;                // True if a live range for this variable has been
+                                               // reported from last call to EndBlock
+
+public:
+    LiveRangeDumper(const LiveRangeList* liveRanges);
+
+    ~LiveRangeDumper(){};
+
+    // Make the dumper point to the last "VariableLiveRange" opened or nullptr if all are closed
+    void resetDumper(const LiveRangeList* list);
+
+    //  Make "LiveRangeDumper" instance points the last "VariableLifeRange" added so we can
+    // starts dumping from there after the actual "BasicBlock"s code is generated.
+    void setDumperStartAt(const LiveRangeListIterator liveRangeIt);
+
+    // Return an iterator to the first "VariableLiveRange" edited/added during the current
+    // "BasicBlock"
+    LiveRangeListIterator getStartForDump() const;
+
+    // Return whether at least a "VariableLiveRange" was alive during the current "BasicBlock"'s
+    // code generation
+    bool hasLiveRangesToDump() const;
+};
+#endif // DEBUG
+
+//--------------------------------------------
+//
+// VariableLiveDescriptor: This class persist and update all the changes
+//  on the variable home of a variable. It has an instance of "LiveRangeList"
+//  and methods to report the start/end of a VariableLiveRange.
+//
+// Notes:
+//  It has an instance of "LiveRangeList"
+//  and methods to report the start/end of a VariableLiveRange.
+//
+class VariableLiveDescriptor
```

Comment:
This (and the dumper too) don't seem to be used outside of `VariableLiveKeeper`. It would be better to move them inside the keeper as it makes it easier to follow their usage.

Reply:
They are inside an #ifdef/#endif to not be included if you are not using VariableLiveRange. What is the goal of doing that? I think nested classes make the bigger one hard to read, but I agree with the uses of nested classes for iterators.

Reply:
I don't see how adding these nested classes make the bigger one harder to read. It's not like their members get mixed with the enclosing class members. As far as readability is concerned, I think it makes it easier to see that these classes are not used outside variable keeper. I only figured that out only after going through all this code.